packages/backend/migrations/20231121184537_add_resource_fkeys_webhook_event.js

```diff
@@ -0,0 +1,81 @@
+exports.up = function (knex) {
+  return knex.schema
+    .table('webhookEvents', function (table) {
+      table
+        .uuid('outgoingPaymentId')
+        .nullable()
+        .references('outgoingPayments.id')
+        .index()
+        .onDelete('CASCADE')
+      table
+        .uuid('incomingPaymentId')
+        .nullable()
+        .references('incomingPayments.id')
+        .index()
+        .onDelete('CASCADE')
+      table
+        .uuid('walletAddressId')
+        .nullable()
+        .references('walletAddresses.id')
+        .index()
+        .onDelete('CASCADE')
+      table
+        .uuid('peerId')
+        .nullable()
+        .references('peers.id')
+        .index()
+        .onDelete('CASCADE')
+      table
+        .uuid('assetId')
+        .nullable()
+        .references('peers.id')
+        .index()
+        .onDelete('CASCADE')
+    })
+    .then(() => {
+      return knex('webhookEvents').update({
+        incomingPaymentId: knex.raw(
+          "CASE WHEN type LIKE 'incoming_payment.%' THEN (data->>'id')::uuid END"
+        ),
+        outgoingPaymentId: knex.raw(
+          "CASE WHEN type LIKE 'outgoing_payment.%' THEN (data->>'id')::uuid END"
+        ),
+        walletAddressId: knex.raw(
+          "CASE WHEN type LIKE 'wallet_address.%' THEN (data->'walletAddress'->>'walletAddressId')::uuid END"
+        ),
+        peerId: knex.raw(
+          "CASE WHEN type LIKE 'peer.%' THEN (data->>'id')::uuid END"
+        ),
+        assetId: knex.raw(
+          "CASE WHEN type LIKE 'asset.%' THEN (data->>'id')::uuid END"
```

Comment:
This updates the existing webhook rows so that one of the resource ids are set. It looks at the start of type (`asset.`, `incoming_payment.`, etc.) to determine the resource type then sets the appropriate id to `data.id` (or `data.walletAddress.walletAddress.id` for event types starting with `wallet_address`).

This is going to come with all sorts of gotchas and it would really be nice to pull in a production dump of rafiki.money and run it against that. Some potential problems:

- there is some unaccounted for event type (either from me not finding it in the code, or one that used to exist and wasnt migrated). In that case this migration will fail when adding the constraint because it will have 0 of the resource ids set.
- if the id is not in `data.id` (or `data.walletAddress.id`). I know the payments used to be `data.payment.id` but I think maybe rafiki was deployed after that?
- if some `payment_pointer` event or the `data.payment_pointer`key was not updated to wallet_address (this should not be the case though... we wrote a migration for this).
- if some resource was hard deleted from the db. the webhook would exist with the data.id but the foreign key column could not be set because the resource no longer exists.

Reply:
Radu pointed out `wallet_address.not_found` wont have a wallet address associated. I think a constraint ensuring <= 1 of these columns is set is the best we can do.

Reply:
the check now ensures exactly 1 is set, unless the event is `wallet_address.not_found`. Additionally, I added checks in the `beforeInsert` hooks to the respective models (`IncomingPaymentEvent`, etc.) ensure the correct id is set. This included making actual `AssetEvent` and `PeerEvent`. Previously we just inserted using `WebhookEvent`.